File: www/core/Polyfills/Workers/SharedWorker.js
importScripts("EventTarget");

class SharedWorker extends EventTarget {

    constructor(scriptURL, name, instance) {
        super();
        if (this.constructor.name !== "SharedWorker") { return; }
        this.instance = instance || scope && scope.createSharedWorker(scriptURL, name);
    }

    get port() {
        return this.instance && this.instance.port.thisJSValue || null;
    }

    static create(instance) {
        return new SharedWorker(instance.scriptURL, instance, instance);
    }

}

File: src/ios/JSCore/Workers/SharedWorker.swift
//
//  SharedWorker.swift
//  WTVJavaScriptCore
//
//  Created by Jing KE on 29/07/16.
//  Copyright © 2016 WizTiVi. All rights reserved.
//

import Foundation
import JavaScriptCore

@objc protocol SharedWorkerJSExport: JSExport {
    
    var scriptURL: String { get }
    var name: String { get }
    
    var port: MessagePort { get }
    
}

class SharedWorker: AbstractWorker, SharedWorkerJSExport {
    
    let scriptURL: String
    var name: String
    
    init(context: ScriptContext, scriptURL: String, name: String = "") {
        self.scriptURL = scriptURL
        self.name = name
        super.init(context: context, scriptURL: scriptURL)
        print("\(self.port)")
    }
    
    override func run() {
        print("==================== [\(self.className)] Start running shared worker ====================")
        SharedWorkerGlobalScope.runWorker(self)
        print("==================== [\(self.className)] End running shared worker ====================")
    }
    
}

extension SharedWorker {
    
    class func create(context: ScriptContext, scriptURL: String, name: String = "") -> SharedWorker {
        return SharedWorker(context: context, scriptURL: scriptURL, name: name)
    }
    
}